File: src/main/scala/com/formious/models/Template.scala
package com.formious.models

import java.time.ZonedDateTime
import java.sql.ResultSet
import com.formious.common.Database.{query, execute}
import com.formious.common.Recoders._

case class Template(id: Int,
                    name: String,
                    html: String,
                    created: ZonedDateTime)

object Template {
  def apply(row: ResultSet) = new Template(
    row.getInt("id"),
    row.getString("name"),
    row.getString("html"),
    row.getTimestamp("created").toZonedDateTime)

  def all = {
    query("SELECT * FROM template ORDER BY id ASC") { Template(_) }
  }

  def find(id: Int) = {
    query("SELECT * FROM template WHERE id = ?", List(id)) { Template(_) }.head
  }

  def create(name: String,
             html: String) = {
    query("""
      INSERT INTO template (name, html)
      VALUES (?, ?) RETURNING *
    """, List(name, html)) { Template(_) }.head
  }

  def update(id: Int,
             name: String,
             html: String) = {
    execute("""
      UPDATE template SET name = ?, html = ?
      WHERE id = ?
    """, List(name, html, id))
  }

  def delete(id: Int) = {
    execute("DELETE FROM template WHERE id = ?", List(id))
  }
}

File: src/com/formious/db/template.clj
(ns com.formious.db.template
  (:require [com.formious.db :as db])
  (:import [java.time ZonedDateTime]))

(defrecord Template [^Integer id ^String name ^String html ^ZonedDateTime created])

(defn row->Template
  [{:keys [id name html created]}]
  (Template. id name html (.toZonedDateTime created)))

(defn blank
  []
  {:name "", :html ""})

(defn all
  []
  (->> (db/query "SELECT * FROM template ORDER BY id ASC") (map row->Template)))

(defn find-by-id
  [id]
  (->> (db/query ["SELECT * FROM template WHERE id = ?" id]) first row->Template))

(defn insert!
  [^String name ^String html]
  (->> (db/insert! "template" [(Template. name html)]) first row->Template))

(defn update!
  [^Integer id ^String name ^String html]
  (db/update! "template" {:name name, :html html} ["id = ?" id]))

(defn delete!
  [^Integer id]
  (db/delete! "template" ["id = ?" id]))
